### DISCUSSION TITLE
Implementing Log4net or Serilog for the logger.

### DISCUSSION THREAD

**Discussion by Elanchezhiyan-P** [artehe/Netimobiledevice#26] [Q&A]
Hi @artehe , 

I tried to implement the logger concept (log4net) for the current build but failed. This is due to multiple concurrency occurring for a single file. Do you have any idea for this process?

**Comment by artehe** [comment]
I'm fairly sure the mainstream loggers such as Log4net and Serilog should be able to handle concurrency pretty well, could you elaborate a bit more on what you were trying to achieve?

Was it for example replacing all the places that currently use `Debug.WriteLine` and `Console.WriteLine` in the actual library with `Log.Information` etc? As providing some logging functionality which is independent of any implementation would probably be quite useful.

**Reply by Elanchezhiyan-P** [reply]
@artehe ,

Yes, implementing the Log4net package helps us in tracing the logs easily. This can be configured to temp folder so that the log files can be cleared easily as well. I tried to implement but the concurrency error occurred. Also, defining the repository is the one of the major error that I faced.

**Reply by artehe** [reply]
I've personally never used log4net myself so can't really offer much help with that.

I did experiment with just using Microsoft.Extensions.Logging and didn't run into any particular issues with writing any logs, although that was just to the console. Wasn't too happy with how I'd implemented it so not gonna add that into the repository unless I can think of a nicer way of doing it which allows you to use whichever logger you want.

Do you have an example you can share of what you tried and I'll try and take a look at it.

**Reply by Elanchezhiyan-P** [reply]
Thanks, @artehe. I will create a new branch in my forked repo and provide the link in the chat so, that you can take a look.

**Reply by artehe** [reply]
Don't know if or how far you've got with this but while doing some other bits, I had another go at it and I'm a lot happier with what I've got now.

You can have a look and see what you think of it if you like in the [logging](https://github.com/artehe/Netimobiledevice/tree/logging) branch

I know it still needs some tweaking to let the log levels correctly and would need to add it to all the other areas of the library as it's currently just implemented for the DeviceBackup class. Should be flexible enough that you can connect whatever logging implementaiton you like to it be that serilog, log4net or others. 

Let me know your thoughts.

**Reply by Elanchezhiyan-P** [reply]
@artehe , I looked at the [logging](https://github.com/artehe/Netimobiledevice/tree/logging) branch and it seems fine. We may need to add logs to a few more class files like BackupStatus, DeviceLink, Mobilebackup2Service, etc which will help us when we create it as an application. Also, if we can write the logs to the text file, that should be more flexible in checking the errors. The path may be in the Temp folder.

**Reply by artehe** [reply]
Cool!

I agree that the logging should be added to more places, I just did that first one as like a proof of concept for what we'd do. Will need to adjust some of the log levels as well as I just set them all to be `LogLevel.Debug` to start with but some will want to be higher and maybe one or two lower. I'll probably go through and do this to start with then might just need a bit of tweaking afterwards.

I don't particularly want the library to be writing these log to a file itself and would be up to you when you use the library to get your logging to do that, as I mentioned there should be enough flexibility there that whatever logging implementation you actually use should be able to handle that we now just expose the actual information that previously we just wrote to Debug using `Debug.WriteLine`.

**Reply by Elanchezhiyan-P** [reply]
That makes sense! Please proceed further and let us work on it. I will update the changes wherever required.

When we were using text files, I was facing the concurrency issue. I will check from my side and let you know.

**Reply by artehe** [reply]
I've added it everywhere else that used to write things out and set the debug levels to what I think they should be at first glance.

There might need to be a little adjustment here and then but have a play round with it and let me know what you think

**Reply by Elanchezhiyan-P** [reply]
@artehe , Sure, Let me check it and get back to you.

**Reply by artehe** [reply]
I've pushed the PR to the main repo now as I wanted to use it, so if there's any changes you want to make to it feel free to make a PR :)

**Reply by Elanchezhiyan-P** [reply]
Thanks! @artehe. Will check now and get back to you
